src/atoms/EditableTitle/EditableTitle.tsx

```diff
@@ -0,0 +1,149 @@
+
+const TitleInput = styled.input`
+  background: ${props => props.theme.background};
+  border: 0.125em solid ${props => props.theme.controlBorder};
+  border-radius: ${borderRadius};
+  color: ${props => props.theme.text};
+  font: ${scale(0.5)} Lato, sans-serif;
+  padding: 0;
+`;
+
+const Content = styled.div`
+  padding-left: ${scale(0)};
+`;
+
+interface BaseAddressLabelProps {
+  title: string;
+}
+
+interface NonEditableAddressLabelProps extends BaseAddressLabelProps {
+  editable: false;
+}
+
+export interface EditableAddressLabelProps extends BaseAddressLabelProps {
+  editable: true;
+  onSubmit(title: string): void;
+}
+
+export type AddressLabelProps =
+  | NonEditableAddressLabelProps
+  | EditableAddressLabelProps;
+
+export interface AddressLabelState {
+  title: string;
+  isTitleEditable: boolean;
+}
+
+const TITLE_PLACEHOLDER = 'John Doe';
+
+export class AddressLabel extends Component<
+  AddressLabelProps,
+  AddressLabelState
+> {
+  public static defaultProps: AddressLabelProps = {
+    title: TITLE_PLACEHOLDER,
+    editable: false,
+  };
+  public state: AddressLabelState;
+
+  constructor(props: AddressLabelProps) {
+    super(props);
+
+    this.state = {
+      isTitleEditable: false,
+      title: props.title,
+    };
+  }
+
+  public handleSubmit = (event: FormEvent) => {
+    event.preventDefault();
+    if (!this.props.editable) {
+      throw Error('HandleSubmit should only be called on an editable address');
+    }
+
+    const { title } = this.state;
+
+    this.setState({ isTitleEditable: false });
+
+    const { onSubmit } = this.props;
+
+    onSubmit(title);
+  };
+
+  public enableTitleEditing = () => {
+    this.setState({ isTitleEditable: true });
+  };
+
+  public handleTitleInput = (event: ChangeEvent<HTMLInputElement>) =>
+    this.setState({ title: event.target.value });
+
+  public renderTitleInput() {
+    return (
+      <form onSubmit={this.handleSubmit}>
+        <TitleInput
+          onClick={this.enableTitleEditing}
+          value={this.state.title}
+          onChange={this.handleTitleInput}
+          autoFocus={true} // eslint-disable-line jsx-a11y/no-autofocus
+        />
+        <SubmitButton />
+      </form>
+    );
+  }
+
+  public renderTitle() {
+    const { title } = this.state;
+    return (
+      <>
+        <Title onClick={this.enableTitleEditing} clickable={true}>
+          {title}
+        </Title>
+        <ColoredIconButton icon="create" onClick={this.enableTitleEditing} />
+      </>
+    );
+  }
+
+  public nonEditableTitle() {
+    return <Title>{this.props.title}</Title>;
+  }
+
+  public editableTitle() {
+    const { isTitleEditable } = this.state;
+
+    return (
+      <Content>
+        {isTitleEditable ? this.renderTitleInput() : this.renderTitle()}
+      </Content>
+    );
+  }
+
+  public render() {
+    const { editable } = this.props;
+    return editable ? this.editableTitle() : this.nonEditableTitle();
+  }
+}
+
+export default AddressLabelProps;
```

Comment:
It seems like this file should be renamed to AddressLabel if we're keeping it separate, though I wouldn't mind putting it in the same file as EditableAddress personally.